File: internal/quota/sub2api_types.go
package quota

import "time"

type Sub2APIAccountUsage struct {
	TotalRequests     int64   `json:"totalRequests"`
	InputTokens       int64   `json:"inputTokens"`
	OutputTokens      int64   `json:"outputTokens"`
	CacheTokens       int64   `json:"cacheTokens"`
	TotalTokens       int64   `json:"totalTokens"`
	ActualCost        float64 `json:"actualCost"`
	AverageDurationMS float64 `json:"averageDurationMs"`
}

type Sub2APIQuotaWindow struct {
	Utilization *float64   `json:"utilization,omitempty"`
	Consumed    int64      `json:"consumed"`
	Limit       *int64     `json:"limit,omitempty"`
	Remaining   *int64     `json:"remaining,omitempty"`
	Ratio       *float64   `json:"ratio,omitempty"`
	WindowStart *time.Time `json:"windowStart,omitempty"`
	WindowEnd   *time.Time `json:"windowEnd,omitempty"`
	RefreshAt   *time.Time `json:"refreshAt,omitempty"`
	Status      string     `json:"status"`
}

type Sub2APIRankingRow struct {
	Dimension     string  `json:"dimension"`
	Name          string  `json:"name"`
	TotalRequests int64   `json:"totalRequests"`
	InputTokens   int64   `json:"inputTokens"`
	OutputTokens  int64   `json:"outputTokens"`
	CacheTokens   int64   `json:"cacheTokens"`
	TotalTokens   int64   `json:"totalTokens"`
	ActualCost    float64 `json:"actualCost"`
	Share         float64 `json:"share"`
}

type Sub2APIRankingTrendPoint struct {
	Bucket string `json:"bucket"`
	Name   string `json:"name"`
	Tokens int64  `json:"tokens"`
}

type Sub2APIEvent struct {
	ID                   int64     `json:"id"`
	CreatedAt            time.Time `json:"createdAt"`
	User                 string    `json:"user"`
	APIKey               string    `json:"apiKey"`
	Model                string    `json:"model"`
	RequestedModel       string    `json:"requestedModel"`
	UpstreamModel        string    `json:"upstreamModel"`
	AccountID            int64     `json:"accountId"`
	AccountName          string    `json:"accountName"`
	Status               string    `json:"status"`
	InputTokens          int64     `json:"inputTokens"`
	OutputTokens         int64     `json:"outputTokens"`
	CacheTokens          int64     `json:"cacheTokens"`
	TotalTokens          int64     `json:"totalTokens"`
	ActualCost           float64   `json:"actualCost"`
	DurationMS           int64     `json:"durationMs"`
	FirstTokenDurationMS *int64    `json:"firstTokenDurationMs,omitempty"`
}

type Sub2APIEventsResponse struct {
	Events []Sub2APIEvent `json:"events"`
	Total  int64          `json:"total"`
	Page   int            `json:"page"`
	Limit  int            `json:"limit"`
}

type Sub2APIAccountQuota struct {
	ID                  int64               `json:"id"`
	Provider            string              `json:"provider"`
	AccountType         string              `json:"accountType"`
	DisplayName         string              `json:"displayName"`
	PlanType            string              `json:"planType,omitempty"`
	Status              string              `json:"status"`
	StatusDetail        string              `json:"statusDetail"`
	StatusResetAt       *time.Time          `json:"statusResetAt,omitempty"`
	HasError            bool                `json:"hasError"`
	Schedulable         bool                `json:"schedulable"`
	SessionWindowStatus string              `json:"sessionWindowStatus,omitempty"`
	ResetAt             *time.Time          `json:"resetAt,omitempty"`
	ExpiresAt           *time.Time          `json:"expiresAt,omitempty"`
	LastUsedAt          *time.Time          `json:"lastUsedAt,omitempty"`
	CredentialKeys      []string            `json:"credentialKeys"`
	Usage               Sub2APIAccountUsage `json:"usage"`
	FiveHourWindow      Sub2APIQuotaWindow  `json:"fiveHourWindow"`
	WeeklyWindow        Sub2APIQuotaWindow  `json:"weeklyWindow"`
}

type Sub2APIOverview struct {
	AccountCount       int64   `json:"accountCount"`
	ActiveAccountCount int64   `json:"activeAccountCount"`
	TotalRequests      int64   `json:"totalRequests"`
	InputTokens        int64   `json:"inputTokens"`
	OutputTokens       int64   `json:"outputTokens"`
	CacheTokens        int64   `json:"cacheTokens"`
	TotalTokens        int64   `json:"totalTokens"`
	ActualCost         float64 `json:"actualCost"`
	AccountCost        float64 `json:"accountCost"`
	ActiveUsers        int64   `json:"activeUsers"`
}

File: web/src/lib/sub2apiTypes.ts
export type Sub2ApiRankingDimension = 'user' | 'api_key' | 'model' | 'account'

export type Sub2ApiQuotaWindow = {
  utilization?: number
  consumed: number
  limit?: number
  remaining?: number
  ratio?: number
  windowStart?: string
  windowEnd?: string
  refreshAt?: string
  status: string
}

export type Sub2ApiRanking = {
  dimension: Sub2ApiRankingDimension
  name: string
  totalRequests: number
  inputTokens: number
  outputTokens: number
  cacheTokens: number
  totalTokens: number
  actualCost: number
  share: number
}

export type Sub2ApiEvent = {
  id: number
  createdAt: string
  user: string
  apiKey: string
  model: string
  requestedModel: string
  upstreamModel: string
  accountId: number
  accountName: string
  status: string
  inputTokens: number
  outputTokens: number
  cacheTokens: number
  totalTokens: number
  actualCost: number
  durationMs: number
  firstTokenDurationMs?: number
}

export type Sub2ApiEventsResponse = {
  events: Sub2ApiEvent[]
  total: number
  page: number
  limit: number
}

export type Sub2ApiAccountUsage = {
  totalRequests: number
  inputTokens: number
  outputTokens: number
  cacheTokens: number
  totalTokens: number
  actualCost: number
  averageDurationMs: number
}

export type Sub2ApiAccount = {
  id: number
  provider: string
  accountType: string
  displayName: string
  planType?: string
  status: string
  statusDetail?: string
  statusResetAt?: string
  hasError?: boolean
  schedulable: boolean
  sessionWindowStatus?: string
  resetAt?: string
  expiresAt?: string
  lastUsedAt?: string
  credentialKeys: string[]
  usage: Sub2ApiAccountUsage
  fiveHourWindow?: Sub2ApiQuotaWindow
  weeklyWindow?: Sub2ApiQuotaWindow
}

export type Sub2ApiOverview = {
  accountCount: number
  activeAccountCount: number
  totalRequests: number
  inputTokens: number
  outputTokens: number
  cacheTokens: number
  totalTokens: number
  actualCost: number
  accountCost: number
  activeUsers: number
}

export type Sub2ApiTimeseriesPoint = {
  bucketStart: string
  totalRequests: number
  inputTokens: number
  outputTokens: number
  cacheCreationTokens: number
  cacheReadTokens: number
  totalCost: number
  actualCost: number
  accountCost: number
  totalDurationMS: number
  activeUsers: number
}

export type Sub2ApiModelUsage = {
  model: string
  requestedModel: string
  upstreamModel: string
  totalRequests: number
  inputTokens: number
  outputTokens: number
  cacheCreationTokens: number
  cacheReadTokens: number
  totalCost: number
  actualCost: number
  averageDurationMS: number
}

export type Sub2ApiHealthBlock = {
  start_time: string
  end_time: string
  success: number
  failure: number
  rate: number
}

export type Sub2ApiServiceHealth = {
  total_success: number
  total_failure: number
  success_rate: number
  rows: number
  columns: number
  bucket_seconds: number
  window_start: string
  window_end: string
  block_details: Sub2ApiHealthBlock[]
}
